sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/IfDirectiveHelper.cs

```diff
@@ -0,0 +1,193 @@
+            return walker.CollectedDirectives;
+        }
+
+        /// <summary>
+        /// Collects all of the #if, #else, #elsif and #endif directives occuring in the
+        /// syntax tree up to the specified node
+        /// </summary>
+        private class BranchingDirectiveCollector : CSharpSyntaxWalker
+        {
+            private readonly SyntaxNode terminatingNode;
+            private bool found;
+            private readonly IList<DirectiveTriviaSyntax> directiveTrivia;
+
+            public BranchingDirectiveCollector(SyntaxNode terminatingNode)
+                :base(SyntaxWalkerDepth.StructuredTrivia)
+            {
+                this.terminatingNode = terminatingNode;
+                found = false;
+                directiveTrivia = new List<DirectiveTriviaSyntax>();
+            }
+
+            public IList<DirectiveTriviaSyntax> CollectedDirectives => directiveTrivia;
+
+            public override void Visit(SyntaxNode node)
+            {
+                // Stop traversing once we've walked down to the terminating node
+                if (found)
+                {
+                    return;
+                }
+
+                if (node == terminatingNode)
+                {
+                    VisitTerminatingNodeLeadingTrivia();
+                    found = true;
+                    return;
+                }
+
+                base.Visit(node);
+            }
+
+            private void VisitTerminatingNodeLeadingTrivia()
+            {
+                // Special case: the leading trivia of the terminating node
+                // could contain directives. However, we won't have processed
+                // these yet, as they are treated as children of the node
+                // even though they appear before it in the text
+                if (terminatingNode.HasLeadingTrivia)
+                {
+                    this.VisitLeadingTrivia(terminatingNode.GetFirstToken(includeZeroWidth: true));
+                }
+            }
+
+            public override void VisitIfDirectiveTrivia(IfDirectiveTriviaSyntax node)
+            {
+                AddDirective(node);
+                base.VisitIfDirectiveTrivia(node);
+            }
+
+            public override void VisitElseDirectiveTrivia(ElseDirectiveTriviaSyntax node)
+            {
+                AddDirective(node);
+                base.VisitElseDirectiveTrivia(node);
+            }
+
+            public override void VisitElifDirectiveTrivia(ElifDirectiveTriviaSyntax node)
+            {
+                AddDirective(node);
+                base.VisitElifDirectiveTrivia(node);
+            }
+
+            public override void VisitEndIfDirectiveTrivia(EndIfDirectiveTriviaSyntax node)
+            {
+                AddDirective(node);
+                base.VisitEndIfDirectiveTrivia(node);
+            }
+
+            private void AddDirective(DirectiveTriviaSyntax node)
+            {
+                if (node.IsActive)
```

Comment:
A syntax node is _active_ if it is in not in an section that is excluded by a conditional compilation block.

```
#if BLOCK1     // BLOCK1 is not defined, but this #if is active
#If BLOCK2     // this #if node is not active since it's in exclude
#endif             // inactive
#endif             // active
```